lib/libvmod_blob/base64.c

```diff
@@ -35,8 +35,6 @@
 
```

Comment:
Responding to the commit message here.

const is more of a hint to programmers than to compilers, so there's no error if it's removed. But why remove the hints? The starting addresses and lengths of the input and output buffers are never changed. The contents of the input buffer are never changed. Arguably const is unnecessary for the non-pointer parameters, but IMO const matters more for the pointer parameters.

My impression from the Flexelint commits in the project is that lint favors the use of const where it is warranted. The criteria may be more than just "qualify with const if you can" -- I don't have access to Flexelint, so I can't see for myself. Maybe someone who does could run Flexelint with and without the const qualifiers -- no need to take out the const's, if Flexelint tells us to put them back in.

restrict does not guarantee non-overlapping regions, nor is it intended to. It's a contract that the caller is required to fulfill, hinting the compiler that it may take advantage of the assumption. It's been years now, but I believe I remember checking that the caller code does indeed satisfy the constraint (however I wouldn't rule out an error).

I don't have numbers about the effect of restrict and don't intend to, it would be too much work. I have gotten in the habit of using restrict when the opportunity presents, after looking at code generated with and without it, and seeing that it can lead to shorter code, because the compiler does not have to account for the case of aliasing. (And after reading reports, with numbers, indicating that it can lead to faster code.)

A consideration with the VMOD was that if it used at all, it may be hot code, churning through memory a byte at a time. Just encoding the 32 bytes in req.hash isn't all that much, but if it's happening in many requests at once, it adds up. If someone uses the trick to implement URL encoding with the VMOD, remember that URLs can be up to kilobytes long. And the VMOD is meant to be general purpose for BLOBs, which can be arbitrarily long.

So IMO a hint to the compiler that it can save some work is worth it. Yes it makes the prototypes longer, and they look a little weird, but that aesthetic is less important.

Reply:
+1 on `const`/`restrict`, IMHO it makes no sense to remove correct attributes with the added benefit of producing more efficient code and implicitly documenting the code.